src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,198 @@
+        return unlink(node(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> x = first; x != null; x = x.next) {
+                if (x.element == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> x = first; x != null; x = x.next) {
+                if (object.equals(x.element)) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void addNodeToStart(T value) {
+
+        Node<T> newNode = new Node<>(first, value, null);
+        first.prev = newNode;
+        first = newNode;
+        size++;
+    }
+
+    private void addNodeToEnd(T value) {
+
+        Node<T> newNode = new Node<>(last, value, null);
+        last.next = newNode;
+        last = newNode;
+        size++;
+    }
+
+    private void checkElementIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds!");
+        }
+    }
+
+    private void checkElementIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index out of bounds!");
+        }
+    }
+
+    private T unlink(Node<T> x) {
+         T element = x.element;
+        Node<T> next = x.next;
+        Node<T> prev = x.prev;
+
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+            x.prev = null;
+        }
+
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            x.next = null;
+        }
+
+        x.element = null;
+        size--;
+        return element;
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        Node<T> currentNode;
+
+        if (index <= size / 2) {
+            currentNode = first;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+        } else {
+            currentNode = last;
+            for (int i = size - 1; i > index; i--) {
+                currentNode = currentNode.prev;
+            }
+        }
+            return currentNode;
+    }
+
+    private Node<T> node(int index) {
+
+        if (index < (size >> 1)) {
+            Node<T> x = first;
+            for (int i = 0; i < index; i++) {
+                x = x.next;
+            }
+            return x;
+        } else {
+            Node<T> x = last;
+            for (int i = size - 1; i > index; i--) {
+                x = x.prev;
+            }
+            return x;
+        }
+    }
+
+    public void addFirstNode(T element) {
```

Comment:
Don't use class Objects. Instead of creating a new `Node` with `null` references for `prev` and `next`, use the existing constructor of `Node` that takes these as parameters.